### PR TITLE
Implements Tags endpoints for Shaarli's REST API

### LINKED ISSUE #904: request for API tag methods
Hello,
I am just wondering what you are thinking of adding method to the API to handle with tags ? There could be different actions like : 

- **get** all tags
- **delete** a tag
- **rename** a tag

### PR DESCRIPTION
**Pull request by ArthurHoaro**
Endpoints:

 * List All Tags [GET]
 * Get a tag [GET]
 * Update a tag [PUT]
 * Delete a tag [DELETE]

Fixes #904
References shaarli/api-documentation#34

Note: I didn't separate endpoints in different commits/PRs because this is *very* similar to `Links` controller.

Also, if we could test this using the Python client before merging it, it would be nice.

### CODE REVIEW

File: application/api/controllers/Tags.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        if (ctype_digit($limit)) {
+            $limit = intval($limit);
+        } elseif ($limit === 'all') {
+            $limit = count($tags);
+        } else {
+            throw new ApiBadParametersException('Invalid limit');
+        }
+
+        $out = [];
+        $cpt = 0;
```
Comment (nodiscc) [inline application/api/controllers/Tags.php]: rename `$cpt` to ` $index` ? Just nitpicking on naming conventions

**Comment by virtualtam** [comment]
> Python client

I'll draft a PR during the week :)

File: application/api/controllers/Tags.php (hunk 1 above)
Comment (ArthurHoaro) [reply application/api/controllers/Tags.php]: Sure!

File: tests/api/controllers/tags/DeleteTagTest.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+        $historyEntry = $this->history->getHistory()[0];
+        $this->assertEquals(\History::UPDATED, $historyEntry['event']);
+        $this->assertTrue(
+            (new \DateTime())->add(\DateInterval::createFromDateString('-5 seconds')) < $historyEntry['datetime']
+        );
+    }
+
+    /**
+     * Test DELETE link endpoint: reach not existing ID.
```
Comment (virtualtam) [inline tests/api/controllers/tags/DeleteTagTest.php]: copy-paste spotted :)

File: tests/api/controllers/tags/GetTagNameTest.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        $response = $this->controller->getTag($request, new Response(), ['tagName' => $tagName]);
+        $this->assertEquals(200, $response->getStatusCode());
+        $data = json_decode((string) $response->getBody(), true);
+        $this->assertEquals(self::NB_FIELDS_TAG, count($data));
+        $this->assertEquals($tagName, $data['name']);
+        $this->assertEquals(2, $data['occurrences']);
+    }
+
+    /**
+     * Test basic getLink service: get non existent link => ApiLinkNotFoundException.
```
Comment (virtualtam) [inline tests/api/controllers/tags/GetTagNameTest.php]: M-w C-y spotted :)

File: tests/api/controllers/tags/GetTagsTest.php [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * After every test, remove the test datastore.
+     */
+    public function tearDown()
+    {
+        @unlink(self::$testDatastore);
+    }
+
+    /**
+     * Test basic getLinks service: returns all tags.
```
Comment (virtualtam) [inline tests/api/controllers/tags/GetTagsTest.php]: C-c C-v spotted :)

**Comment by virtualtam** [comment]
With Apache HTTPD 2.4, I added the following settings to `.htaccess` for the `PUT` and `DELETE` verbs to be accepted (otherwise resulting in a `403 Access forbidden`:

```apache
[...]

# REST API
<Limit GET POST PUT DELETE OPTIONS>
    Require all granted
</Limit>
<LimitExcept GET POST PUT DELETE OPTIONS>
    Require all denied
</LimitExcept>

[...]
```

Resources:
- https://stackoverflow.com/questions/25696539/put-request-403-forbidden
- https://serverfault.com/questions/275512/put-request-results-in-403-forbidden-need-apache-to-allow-put-requests
- https://httpd.apache.org/docs/2.4/en/mod/core.html#limit

I'm not sure this is the cleanest way to do this, but at least it allowed me to test the new endpoints through the Python client :)

Also, the client currently has no support for `204 No content` returned when deleting resources, we could do one of the following:
- let the API return a `200` with a deletion message
- add support for `204` in the client

**Comment by ArthurHoaro** [comment]
That's a good point, thank you I'll add these to `.htaccess`. It should also be documented in the Apache configuration.

Also, I wrote a browser extension recently, unrelated to Shaarli, but we should add [CORS](https://developer.mozilla.org/fr/docs/Web/HTTP/CORS) headers in our API response, otherwise I don't think AJAX request can work. I'll raise a separate issue.

> add support for 204 in the client

That's the expected behaviour, so I guess the client should support it.

**Comment by ArthurHoaro** [comment]
Updated

**Comment by ArthurHoaro** [comment]
> I'll raise a separate issue.

#1174

**Pull request merged**